*** src/include/pcp/pcp_stream.h	2021-08-04 02:31:46.106372420 +0000
--- pcp_stream.h	2021-08-04 02:32:22.082659482 +0000
***************
*** 49,54 ****
  extern int	pcp_write(PCP_CONNECTION * pc, void *buf, int len);
  extern int	pcp_flush(PCP_CONNECTION * pc);
  
! #define UNIX_DOMAIN_PATH "/tmp"
  
  #endif							/* PCP_STREAM_H */
--- 49,54 ----
  extern int	pcp_write(PCP_CONNECTION * pc, void *buf, int len);
  extern int	pcp_flush(PCP_CONNECTION * pc);
  
! #define UNIX_DOMAIN_PATH "/var/run/postgresql"
  
  #endif							/* PCP_STREAM_H */
