*** src/sample/pgpool.conf.sample	2023-11-07 20:38:38.890984731 +0900
--- pgpool.conf.sample	2023-11-09 13:22:19.179462724 +0900
***************
*** 40,46 ****
  #port = 9999
                                     # Port number
                                     # (change requires restart)
! #unix_socket_directories = '/tmp'
                                     # Unix domain socket path(s)
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 40,46 ----
  #port = 9999
                                     # Port number
                                     # (change requires restart)
! unix_socket_directories = '/var/run/postgresql'
                                     # Unix domain socket path(s)
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 67,73 ****
  #pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! #pcp_socket_dir = '/tmp'
                                     # Unix domain socket path(s) for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 67,73 ----
  #pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path(s) for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 736,742 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! #wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 736,742 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
