*** src/sample/pgpool.conf.sample	2025-09-30 17:44:36.223222257 +0900
--- pgpool.conf.sample	2025-09-30 17:46:22.952860803 +0900
***************
*** 82,96 ****

  # - Backend Connection Settings -

! #backend_hostname0 = 'host1'
                                     # Host name or IP address to connect to for backend 0
! #backend_port0 = 5432
                                     # Port number for backend 0
! #backend_weight0 = 1
                                     # Weight for backend 0 (only in load balancing mode)
! #backend_data_directory0 = '/data'
                                     # Data directory for backend 0
! #backend_flag0 = 'ALLOW_TO_FAILOVER'
                                     # Controls various backend behavior
                                     # ALLOW_TO_FAILOVER, DISALLOW_TO_FAILOVER
                                     # or ALWAYS_PRIMARY
--- 82,96 ----

  # - Backend Connection Settings -

! backend_hostname0 = 'localhost'
                                     # Host name or IP address to connect to for backend 0
! backend_port0 = 5432
                                     # Port number for backend 0
! backend_weight0 = 1
                                     # Weight for backend 0 (only in load balancing mode)
! backend_data_directory0 = '/var/lib/pgsql/data'
                                     # Data directory for backend 0
! backend_flag0 = 'ALLOW_TO_FAILOVER'
                                     # Controls various backend behavior
                                     # ALLOW_TO_FAILOVER, DISALLOW_TO_FAILOVER
                                     # or ALWAYS_PRIMARY
***************
*** 351,357 ****
                                     # location of pgpool.conf file or
                                     # as an absolute path
                                     # (change requires restart)
! #work_dir = '/tmp'
                                     # Directory for pgpool_status and lock files.
                                     # (change requires restart)

--- 351,357 ----
                                     # location of pgpool.conf file or
                                     # as an absolute path
                                     # (change requires restart)
! work_dir = '/var/lib/pgpool'
                                     # Directory for pgpool_status and lock files.
                                     # (change requires restart)
  
