*** src/sample/pgpool.conf.sample	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample	2022-05-18 15:21:51.385047148 +0900
***************
*** 39,45 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 39,45 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 59,65 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 59,65 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 665,671 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 665,671 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
*** src/sample/pgpool.conf.sample-stream	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample-stream	2022-05-18 15:20:21.918452223 +0900
***************
*** 39,45 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 39,45 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 59,65 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 59,65 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 665,671 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 665,671 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
*** src/sample/pgpool.conf.sample-replication	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample-replication	2022-05-18 15:21:17.457821544 +0900
***************
*** 39,45 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 39,45 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 65,71 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 65,71 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 663,669 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 663,669 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
*** src/sample/pgpool.conf.sample-logical	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample-logical	2022-05-18 15:20:59.437701716 +0900
***************
*** 38,44 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 38,44 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 58,64 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 58,64 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 627,633 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 627,633 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
*** src/sample/pgpool.conf.sample-raw	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample-raw	2022-05-18 15:20:37.478555694 +0900
***************
*** 39,45 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 39,45 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 59,65 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 59,65 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 666,672 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 666,672 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
*** src/sample/pgpool.conf.sample-slony	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample-slony	2022-05-18 15:20:42.323587911 +0900
***************
*** 39,45 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 39,45 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 65,71 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 65,71 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 664,670 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 664,670 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
*** src/sample/pgpool.conf.sample-snapshot	2022-05-18 15:22:35.044337469 +0900
--- pgpool.conf.sample-snapshot	2022-05-18 15:20:47.797624309 +0900
***************
*** 39,45 ****
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/tmp'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 39,45 ----
  port = 9999
                                     # Port number
                                     # (change requires restart)
! socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 65,71 ****
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/tmp'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
--- 65,71 ----
  pcp_port = 9898
                                     # Port number for pcp
                                     # (change requires restart)
! pcp_socket_dir = '/var/run/postgresql'
                                     # Unix domain socket path for pcp
                                     # The Debian package defaults to
                                     # /var/run/postgresql
***************
*** 661,667 ****
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/tmp'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
--- 661,667 ----
                                      # Authentication key for watchdog communication
                                      # (change requires restart)
  
! wd_ipc_socket_dir = '/var/run/postgresql'
                                      # Unix domain socket path for watchdog IPC socket
                                      # The Debian package defaults to
                                      # /var/run/postgresql
